adam-core/src/main/scala/org/bdgenomics/adam/rdd/GenomicRDD.scala

```diff
@@ -916,28 +1022,125 @@ trait GenomicRDD[T, U <: GenomicRDD[T, U]] {
       xTag: ClassTag[X],
       otixTag: ClassTag[(Option[T], Iterable[X])]): GenomicRDD[(Option[T], Iterable[X]), Z] = RightOuterShuffleJoinAndGroupByLeft.time {
 
-    val (partitionSize, endSequences) = joinPartitionSizeAndSequences(optPartitions,
-      genomicRdd)
+    val (leftRddToJoin, rightRddToJoin) =
+      prepareForShuffleRegionJoin(genomicRdd, optPartitions)
+
+    // what sequences do we wind up with at the end?
+    val combinedSequences = sequences ++ genomicRdd.sequences
 
-    // key the RDDs and join
     GenericGenomicRDD[(Option[T], Iterable[X])](
-      RightOuterShuffleRegionJoinAndGroupByLeft[T, X](endSequences,
-        partitionSize,
-        rdd.context).partitionAndJoin(flattenRddByRegions(),
-          genomicRdd.flattenRddByRegions()),
-      endSequences,
+      RightOuterShuffleRegionJoinAndGroupByLeft[T, X](leftRddToJoin, rightRddToJoin)
+        .compute(),
+      combinedSequences,
       kv => {
         (kv._2.flatMap(v => genomicRdd.getReferenceRegions(v)) ++
           kv._1.toSeq.flatMap(v => getReferenceRegions(v))).toSeq
+      }).asInstanceOf[GenomicRDD[(Option[T], Iterable[X]), Z]]
+  }
+
+  /**
+   * Copartitions two RDDs according to their ReferenceRegions.
+   *
+   * @note This is best used under the condition that (repeatedly)
+   *   repartitioning is more expensive than calculating the proper location
+   *   of the records of this.rdd. It requires a pass through the co-located
+   *   RDD to get the correct partition(s) for each record. It will assign a
+   *   record to multiple partitions if necessary.
+   *
+   * @param rddToCoPartitionWith The rdd to copartition to.
+   * @return The newly repartitioned rdd.
+   */
+  private[rdd] def copartitionByReferenceRegion[X, Y <: GenomicRDD[X, Y]](
+    rddToCoPartitionWith: GenomicRDD[X, Y])(implicit tTag: ClassTag[T], xTag: ClassTag[X]): U = {
+
+    // if the other RDD is not sorted, we can't guarantee proper copartition
+    assert(rddToCoPartitionWith.isSorted,
+      "Cannot copartition with an unsorted rdd!")
+
+    val destinationPartitionMap = rddToCoPartitionWith.optPartitionMap.get
+    //number of partitions we will have after repartition
```

Comment:
Nits for the code in this function:

* there should be whitespace before a comment
* space after `//`